lib/pure/endians.nim

```diff
@@ -98,16 +126,919 @@ else:
+  ## location. The type of the number must be passed in as an argument.
+  runnableExamples:
+    var buf: array[4, byte]
+    buf[0] = 0xde
+    buf[1] = 0xad
+    buf[2] = 0xbe
+    buf[3] = 0xef
+    doAssert fromBytesBE(int32, buf[0].addr) == 0xdeadbeef'i32
+
+  let a = cast[ptr T](buf)[]
+  when system.cpuEndian == bigEndian: a
+  else: swapEndian(a)
+
+
+func slowFromBytes(T: typedesc[SomeNumber], srcEndian: Endianness,
+                   buf: openArray[byte], startIndex: Natural): T =
+  template fromBytes(S: typedesc[SomeInteger]): T =
+    var a = S(0)
+    for i in 0..<sizeof(T):
+      let s = if srcEndian == bigEndian: (sizeof(T) - i - 1) * 8
+              else: i * 8
+      a = a or (S(buf[startIndex + i]) shl s)
+    cast[T](a)
+
+  when sizeof(T) == 1: result = fromBytes(uint8)
+  elif sizeof(T) == 2: result = fromBytes(uint16)
+  elif sizeof(T) == 4: result = fromBytes(uint32)
+  elif sizeof(T) == 8: result = fromBytes(uint64)
+
+
+func fromBytesBE*(T: typedesc[SomeNumber], buf: openArray[byte],
+                  startIndex: Natural): T {.inline, since: (1,1).} =
+  ## Reads a number represented as a big-endian byte sequence from `buf`
+  ## starting at `startIndex`. The type of the number must be passed in as an
+  ## argument.
+  runnableExamples:
+    var buf: array[6, byte]
+    buf[1] = 0xde
+    buf[2] = 0xad
+    buf[3] = 0xbe
+    buf[4] = 0xef
+    doAssert fromBytesBE(int32, buf, 1) == 0xdeadbeef'i32
+
+  when nimvm: slowFromBytes(T, bigEndian, buf, startIndex)
+  else:
+    assert startIndex < buf.len
+    assert startIndex + sizeof(T) <= buf.len
+    fromBytesBE(T, buf[startIndex].unsafeAddr)
+
+func fromBytesLE*(T: typedesc[SomeNumber],
+                  buf: pointer): T {.inline, since: (1,1).} =
+  ## Reads a number represented as a little-endian byte sequence from a memory
+  ## buffer. The type of the number must be passed in as an argument.
+  runnableExamples:
+    var buf: array[6, uint8]
+    buf[1] = 0xef
+    buf[2] = 0xbe
+    buf[3] = 0xad
+    buf[4] = 0xde
+    doAssert fromBytesLE(int32, buf[1].addr) == 0xdeadbeef'i32
+
+  let a = cast[ptr T](buf)[]
+  when system.cpuEndian == littleEndian: a
+  else: swapEndian(a)
+
+
+func fromBytesLE*(T: typedesc[SomeNumber], buf: openArray[byte],
+                  startIndex: Natural): T {.inline, since: (1,1).} =
+  ## Reads a number represented as a little-endian byte sequence from `buf`
+  ## starting at `startIndex`. The type of the number must be passed in as an
+  ## argument.
+  runnableExamples:
+    var buf: array[6, byte]
+    buf[1] = 0xef
+    buf[2] = 0xbe
+    buf[3] = 0xad
+    buf[4] = 0xde
+    doAssert fromBytesLE(int32, buf, 1) == 0xdeadbeef'i32
+
+  when nimvm: slowFromBytes(T, littleEndian, buf, startIndex)
+  else:
+    assert startIndex < buf.len
+    assert startIndex + sizeof(T) <= buf.len
+    fromBytesLE(T, buf[startIndex].unsafeAddr)
+
+
+when defined(testing) and isMainModule:
```

Comment:
Why do you check on `defined(testing)`? As far as I am aware a `when isMainModule` is enough.
Apart from that. The pattern that when I want to test something on the VM and at runtime, I run the exact same code at compile time and at runtime:

```nim
when isMainModule:
  proc localTest() = [...] # can be a template, too
  localTest()
  static:
    localTest()
```

Reply:
Good points, I've done these changes.